File: sdk/java/src/main/java/dev/udb/client/UdbProject.java
package dev.udb.client;

import com.udb.core.analytics.services.v1.AnalyticsServiceGrpc;
import com.udb.core.apikey.services.v1.ApiKeyServiceGrpc;
import com.udb.core.apikey.services.v1.CreateApiKeyRequest;
import com.udb.core.apikey.services.v1.CreateApiKeyResponse;
import com.udb.core.apikey.services.v1.RevokeApiKeyRequest;
import com.udb.core.apikey.services.v1.RevokeApiKeyResponse;
import com.udb.core.authn.services.v1.AuthnResponse;
import com.udb.core.authn.services.v1.LoginRequest;
import com.udb.core.authn.services.v1.LoginResponse;
import com.udb.core.authn.services.v1.Principal;
import com.udb.core.notification.services.v1.NotificationServiceGrpc;
import com.udb.core.notification.services.v1.SendNotificationRequest;
import com.udb.core.notification.services.v1.SendNotificationResponse;
import com.udb.core.tenant.services.v1.CreateTenantRequest;
import com.udb.core.tenant.services.v1.CreateTenantResponse;
import com.udb.core.tenant.services.v1.TenantServiceGrpc;
import io.grpc.ManagedChannel;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * One-stop project facade over the UDB control- and data-plane services. It
 * shares a single {@link UdbMetadata} identity across all sub-clients, attaches
 * the caller headers to every generated stub, and owns the channels it creates
 * (data-plane + auth-plane), closing them via {@link #close()}.
 *
 * <p>Exposed sub-clients: {@link #data()}, {@link #auth()}/{@link #authz()},
 * {@link #apiKey()}, {@link #tenant()}, {@link #notification()},
 * {@link #analytics()}, {@link #storage()}, {@link #asset()}, and
 * {@link #webRtc()}. The storage/asset/webrtc services live on the native
 * control-plane listener (same target as auth), so they reuse the existing
 * control-plane channel.
 */
public final class UdbProject implements AutoCloseable {
  private final UdbProjectConfig config;
  private final UdbMetadataRef metadata;

  private final ManagedChannel dataChannel;
  private final ManagedChannel authChannel; // may be the same as dataChannel
  private final ManagedChannel webrtcChannel; // may be the same as authChannel

  /** Shared, mutable credentials sent on every outbound call. Mutating it
   *  (via {@link #setCredentials}) hot-swaps the bearer/API key on all stubs. */
  private final UdbCredentials credentials;

  private final UdbClient data;
  private final UdbAuthClient auth;
  private final ApiKeyServiceGrpc.ApiKeyServiceBlockingStub apiKey;
  private final TenantServiceGrpc.TenantServiceBlockingStub tenant;
  private final NotificationServiceGrpc.NotificationServiceBlockingStub notification;
  private final AnalyticsServiceGrpc.AnalyticsServiceBlockingStub analytics;
  private final UdbStorageClient storage;
  private final UdbAssetClient asset;
  private final UdbWebRtcClient webRtc;

  private UdbProject(UdbProjectConfig config) {
    this.config = config;
    UdbMetadata initialMetadata = config.metadata();
    this.metadata = new UdbMetadataRef(initialMetadata);
    this.credentials = UdbCredentials.fromMetadata(initialMetadata);

    this.dataChannel = channel(config.target(), config.tls());
    this.authChannel =
        config.authTarget().equals(config.target())
            ? dataChannel
            : channel(config.authTarget(), config.tls());
    // WebRTC rides the control-plane channel unless a distinct webrtcTarget is
    // configured.
    if (config.webrtcTarget().equals(config.authTarget())) {
      this.webrtcChannel = authChannel;
    } else if (config.webrtcTarget().equals(config.target())) {
      this.webrtcChannel = dataChannel;
    } else {
      this.webrtcChannel = channel(config.webrtcTarget(), config.tls());
    }

    this.data = new UdbClient(dataChannel, metadata, credentials);
    this.auth = new UdbAuthClient(authChannel, metadata, credentials);

    // Native control-plane stubs hang off the auth/control-plane channel; the
    // dynamic credential interceptor attaches shared identity headers + live
    // credentials.
    io.grpc.ClientInterceptor creds = UdbClient.credentialInterceptor(metadata, credentials);
    this.apiKey = ApiKeyServiceGrpc.newBlockingStub(authChannel).withInterceptors(creds);
    this.tenant = TenantServiceGrpc.newBlockingStub(authChannel).withInterceptors(creds);
    this.notification =
        NotificationServiceGrpc.newBlockingStub(authChannel).withInterceptors(creds);
    this.analytics = AnalyticsServiceGrpc.newBlockingStub(authChannel).withInterceptors(creds);

    // Storage / asset are served on the native control-plane listener (same target
    // as auth); WebRTC rides its own channel when webrtcTarget is distinct.
    this.storage = new UdbStorageClient(authChannel, metadata, credentials);
    this.asset = new UdbAssetClient(authChannel, metadata, credentials);
    this.webRtc = new UdbWebRtcClient(webrtcChannel, metadata, credentials);
  }

  /** Open a project facade from config. See {@link Udb#project(UdbProjectConfig)}. */
  public static UdbProject open(UdbProjectConfig config) {
    return new UdbProject(config);
  }

  private static ManagedChannel channel(String target, boolean tls) {
    // Long-lived channel with UDB keepalive + UNAVAILABLE retry; reused across RPCs.
    return UdbChannels.forTarget(target, tls);
  }

  // ── Raw clients / stubs (reachable, never hidden) ──────────────────────────
  public UdbClient data() {
    return data;
  }

  /** Auth ergonomics (authn + authz helpers, cache-friendly). */
  public UdbAuthClient auth() {
    return auth;
  }

  /** Alias for {@link #auth()} — the authorization surface lives on the same client. */
  public UdbAuthClient authz() {
    return auth;
  }

  public ApiKeyServiceGrpc.ApiKeyServiceBlockingStub apiKey() {
    return apiKey;
  }

  public TenantServiceGrpc.TenantServiceBlockingStub tenant() {
    return tenant;
  }

  public NotificationServiceGrpc.NotificationServiceBlockingStub notification() {
    return notification;
  }

  public AnalyticsServiceGrpc.AnalyticsServiceBlockingStub analytics() {
    return analytics;
  }

  /** Native file storage (presigned upload/download, file metadata CRUD). */
  public UdbStorageClient storage() {
    return storage;
  }

  /** Native asset registration + processing pipelines. */
  public UdbAssetClient asset() {
    return asset;
  }

  /** Native WebRTC: room/peer/track/turn accessors + async signaling stream. */
  public UdbWebRtcClient webRtc() {
    return webRtc;
  }

  public UdbProjectConfig config() {
    return config;
  }

  public UdbMetadata metadata() {
    return metadata.current();
  }

  public UdbEntityHandle entity(String messageType, String... key) {
    return data.entity(messageType, key);
  }

  public UdbEntityHandle table(String name, String... key) {
    return data.table(name, key);
  }

  /** The shared credentials holder; mutating it hot-swaps creds on every stub. */
  public UdbCredentials credentials() {
    return credentials;
  }

  /**
   * Hot-swap the bearer token and API key sent on every subsequent call across
   * all sub-clients (data, auth, control-plane, storage, asset, WebRTC). Call
   * this after a login/refresh so the new token reaches outbound metadata — the
   * Java analogue of the TypeScript {@code core.setCredentials}.
   */
  public void setCredentials(String bearerToken, String apiKey) {
    this.credentials.set(bearerToken, apiKey);
  }

  /** Hot-swap just the bearer token (keeps the configured API key). */
  public void setBearerToken(String bearerToken) {
    this.credentials.setBearerToken(bearerToken);
  }

  // ── Convenience wrappers ───────────────────────────────────────────────────

  public AuthnResponse loginAndAdoptTenant(String username, String password) {
    return loginAndAdoptTenant(
        LoginRequest.newBuilder()
            .setUsername(username == null ? "" : username)
            .setPassword(password == null ? "" : password)
            .build());
  }

  /**
   * Login, ALWAYS authenticate the freshly minted bearer, then adopt the
   * canonical tenant/project from the verified principal across every facade.
   */
  public synchronized AuthnResponse loginAndAdoptTenant(LoginRequest request) {
    LoginResponse login = auth.login(request);
    String token = login.getAccessToken().isBlank() ? login.getSessionToken() : login.getAccessToken();
    if (token.isBlank()) {
      throw new IllegalStateException(
          "udb: Login returned no access token (MFA required: " + login.getMfaRequired() + ")");
    }

    AuthnResponse verified = auth.authenticateBearer(token);
    if (!verified.hasPrincipal()) {
      throw new IllegalStateException("udb: AuthenticateBearer returned no principal");
    }

    Principal principal = verified.getPrincipal();
    UdbMetadata current = metadata.current();
    UdbMetadata adopted =
        new UdbMetadata(
            principal.getTenantId().isBlank() ? current.tenantId() : principal.getTenantId(),
            current.purpose(),
            current.correlationId(),
            current.scopes(),
            current.serviceIdentity(),
            principal.getUserId().isBlank() ? current.userId() : principal.getUserId(),
            principal.getProjectId().isBlank() ? current.projectId() : principal.getProjectId(),
            current.clientCatalogVersion(),
            token,
            credentials.apiKey(),
            current.consistency(),
            current.primaryRead(),
            current.maxReplicaLagMs(),
            current.eventualConsistencyAllowed(),
            current.readFenceJson());
    metadata.set(adopted);
    credentials.setBearerToken(token);
    return verified;
  }

  /** Send a notification to a recipient over one or more channels. */
  public SendNotificationResponse sendNotification(SendNotificationRequest request) {
    return notification.sendNotification(request);
  }

  /** Mint a new API key for an owner, returning the response (plain key once). */
  public CreateApiKeyResponse createApiKey(String name, String ownerId, List<String> scopes) {
    return apiKey.createApiKey(CreateApiKeyRequest.newBuilder()
        .setName(name)
        .setOwnerId(ownerId)
        .addAllScopes(scopes == null ? List.of() : scopes)
        .build());
  }

  /**
   * Revoke an API key by id. The ApiKey service exposes create + revoke (no
   * dedicated rotate RPC); rotation is "revoke old, then create new".
   */
  public RevokeApiKeyResponse revokeApiKey(String keyId, String reason) {
    return apiKey.revokeApiKey(RevokeApiKeyRequest.newBuilder()
        .setKeyId(keyId)
        .setRevokeReason(reason == null ? "" : reason)
        .build());
  }

  /**
   * Revoke an existing key and mint a replacement with the same name + scopes —
   * the SDK-side "rotate" convenience over the create/revoke RPCs.
   */
  public CreateApiKeyResponse rotateApiKey(
      String oldKeyId, String name, String ownerId, List<String> scopes) {
    revokeApiKey(oldKeyId, "rotated");
    return createApiKey(name, ownerId, scopes);
  }

  /** Onboard / create a tenant by code + display name. */
  public CreateTenantResponse createTenant(String code, String name) {
    return tenant.createTenant(CreateTenantRequest.newBuilder()
        .setCode(code)
        .setName(name)
        .build());
  }

  @Override
  public void close() {
    shutdown(dataChannel);
    if (authChannel != dataChannel) {
      shutdown(authChannel);
    }
    if (webrtcChannel != dataChannel && webrtcChannel != authChannel) {
      shutdown(webrtcChannel);
    }
  }

  private static void shutdown(ManagedChannel channel) {
    channel.shutdown();
    try {
      if (!channel.awaitTermination(5, TimeUnit.SECONDS)) {
        channel.shutdownNow();
      }
    } catch (InterruptedException err) {
      channel.shutdownNow();
      Thread.currentThread().interrupt();
    }
  }
}

File: sdk/csharp/Udb.Client/UdbProject.cs
using Grpc.Net.Client;
using Udb.Entity.V1;
using Udb.Services.V1;
using AnalyticsV1 = udb.core.Analytics.Services.V1;
using ApikeyV1 = udb.core.Apikey.Services.V1;
using AssetV1 = udb.core.Asset.Services.V1;
using NotificationV1 = udb.core.Notification.Services.V1;
using StorageV1 = udb.core.Storage.Services.V1;
using TenantV1 = udb.core.Tenant.Services.V1;
using WebRtcV1 = udb.core.Webrtc.Services.V1;

namespace Udb.Client;

/// <summary>
/// One configuration object for a <see cref="UdbProject"/>. All fields have
/// sensible defaults; only <see cref="Target"/> is required. <see cref="AuthTarget"/>
/// defaults to <see cref="Target"/> when unset (a single broker fronts both the
/// data plane and the native auth services in the integration stack).
/// </summary>
public sealed record UdbProjectConfig
{
    /// <summary>Data-plane / default gRPC endpoint, e.g. <c>http://localhost:50051</c>.</summary>
    public required string Target { get; init; }

    /// <summary>Auth-plane endpoint. Defaults to <see cref="Target"/>.</summary>
    public string? AuthTarget { get; init; }

    /// <summary>
    /// WebRTC signalling/peer endpoint. Defaults to the (resolved)
    /// <see cref="AuthTarget"/> — the native control-plane listener that serves
    /// the WebRTC services. When set and distinct, the <see cref="UdbProject.WebRtc"/>
    /// facade dials a dedicated channel.
    /// </summary>
    public string? WebrtcTarget { get; init; }

    public string TenantId { get; init; } = "default";
    public string ProjectId { get; init; } = "default";
    public string Purpose { get; init; } = "";
    public string UserId { get; init; } = "";
    public string ServiceIdentity { get; init; } = "";
    public string CorrelationId { get; init; } = "";
    public string[] Scopes { get; init; } = Array.Empty<string>();
    public string ClientCatalogVersion { get; init; } = UdbClient.ProtocolVersion;
    public string BearerToken { get; init; } = "";
    public string ApiKey { get; init; } = "";

    /// <summary>Default TTL for the shared authz decision cache.</summary>
    public TimeSpan AuthzCacheTtl { get; init; } = TimeSpan.FromSeconds(30);

    /// <summary>
    /// Shared secret for HMAC-SHA256 verification of signed policy bundles.
    /// When set, <see cref="UdbProject.Authz"/>.GetPolicyBundleAsync verifies
    /// every returned bundle and throws on mismatch. Unset disables verification.
    /// </summary>
    public string? PolicyBundleSecret { get; init; }
}

/// <summary>
/// Top-level ergonomic facade over every native UDB service whose generated
/// client ships in this package. Mirrors the Go/Python/TS <c>Project</c> facade:
/// one config, shared metadata + channels, and first-class accessors. Raw
/// generated clients stay reachable (<see cref="ApiKey"/>, <see cref="Tenant"/>,
/// …) for RPCs not yet wrapped.
/// The native media services — <see cref="Storage"/>, <see cref="Asset"/>, and
/// <see cref="WebRtc"/> — share the control-plane channel (the same target as
/// auth) and the shared metadata headers.
/// </summary>
public sealed class UdbProject : IAsyncDisposable, IDisposable
{
    private readonly GrpcChannel _dataChannel;
    private readonly GrpcChannel _authChannel;
    private readonly GrpcChannel _nativeServicesChannel;
    private readonly GrpcChannel _webrtcChannel;
    private readonly bool _separateAuthChannel;
    private readonly bool _separateWebrtcChannel;
    private readonly UdbMetadata _metadata;
    private readonly UdbCredentials _credentials;

    private UdbProject(UdbProjectConfig config)
    {
        _metadata = new UdbMetadata(
            TenantId: config.TenantId,
            Purpose: config.Purpose,
            CorrelationId: string.IsNullOrEmpty(config.CorrelationId) ? Guid.NewGuid().ToString("N") : config.CorrelationId,
            Scopes: config.Scopes,
            ServiceIdentity: config.ServiceIdentity,
            UserId: config.UserId,
            ProjectId: config.ProjectId,
            ClientCatalogVersion: config.ClientCatalogVersion,
            BearerToken: config.BearerToken,
            ApiKey: config.ApiKey);
        // Shared, mutable credentials sent on every outbound call. SetCredentials
        // mutates this holder so a refreshed token reaches all clients at once.
        _credentials = new UdbCredentials(config.BearerToken, config.ApiKey);

        var authTarget = string.IsNullOrEmpty(config.AuthTarget) ? config.Target : config.AuthTarget!;
        var webrtcTarget = string.IsNullOrEmpty(config.WebrtcTarget) ? authTarget : config.WebrtcTarget!;
        _dataChannel = GrpcChannel.ForAddress(config.Target);
        _separateAuthChannel = !string.Equals(authTarget, config.Target, StringComparison.Ordinal);
        _authChannel = _separateAuthChannel ? GrpcChannel.ForAddress(authTarget) : _dataChannel;
        _nativeServicesChannel = _authChannel;
        // WebRTC rides the control-plane channel unless a distinct webrtcTarget is set.
        if (string.Equals(webrtcTarget, authTarget, StringComparison.Ordinal))
        {
            _webrtcChannel = _authChannel;
            _separateWebrtcChannel = false;
        }
        else if (string.Equals(webrtcTarget, config.Target, StringComparison.Ordinal))
        {
            _webrtcChannel = _dataChannel;
            _separateWebrtcChannel = false;
        }
        else
        {
            _webrtcChannel = GrpcChannel.ForAddress(webrtcTarget);
            _separateWebrtcChannel = true;
        }

        var cache = new AuthzCache(config.AuthzCacheTtl);

        Data = new DataBroker.DataBrokerClient(_dataChannel);
        Auth = new UdbAuthClient(_authChannel, _metadata, cache, config.PolicyBundleSecret, _credentials);
        Authz = Auth; // same wrapper exposes the authz surface
        ApiKey = new ApikeyV1.ApiKeyService.ApiKeyServiceClient(_nativeServicesChannel);
        Tenant = new TenantV1.TenantService.TenantServiceClient(_nativeServicesChannel);
        Notification = new NotificationV1.NotificationService.NotificationServiceClient(_nativeServicesChannel);
        Analytics = new AnalyticsV1.AnalyticsService.AnalyticsServiceClient(_nativeServicesChannel);

        // Native media services live on the control-plane listener (same target as
        // auth) and reuse the shared metadata headers; WebRTC rides its own channel
        // when webrtcTarget is distinct.
        Storage = new UdbStorageClient(
            new StorageV1.StorageService.StorageServiceClient(_nativeServicesChannel), Headers);
        Asset = new UdbAssetClient(
            new AssetV1.AssetService.AssetServiceClient(_nativeServicesChannel), Headers);
        WebRtc = new UdbWebRtcClient(
            new WebRtcV1.RoomService.RoomServiceClient(_webrtcChannel),
            new WebRtcV1.PeerService.PeerServiceClient(_webrtcChannel),
            new WebRtcV1.TrackService.TrackServiceClient(_webrtcChannel),
            new WebRtcV1.TurnService.TurnServiceClient(_webrtcChannel),
            new WebRtcV1.SignalingService.SignalingServiceClient(_webrtcChannel),
            Headers);
    }

    /// <summary>Constructs a project facade. Async for symmetry with the other SDKs.</summary>
    public static Task<UdbProject> ProjectAsync(UdbProjectConfig config)
        => Task.FromResult(new UdbProject(config));

    /// <summary>Synchronous constructor for callers that don't need the async form.</summary>
    public static UdbProject Open(UdbProjectConfig config) => new(config);

    /// <summary>Shared metadata headers (eight wire/identity headers).</summary>
    public Grpc.Core.Metadata Headers()
    {
        var headers = new Grpc.Core.Metadata
        {
            { "x-tenant-id", _metadata.TenantId },
            { "x-user-id", _metadata.UserId },
            { "x-purpose", _metadata.Purpose },
            { "x-correlation-id", _metadata.CorrelationId },
            { "x-scopes", string.Join(",", _metadata.Scopes) },
            { "x-service-identity", _metadata.ServiceIdentity },
            { "x-udb-project-id", _metadata.ProjectId },
            { "x-udb-client-catalog-version", _metadata.ClientCatalogVersion },
        };
        var bearer = _credentials.BearerToken;
        if (!string.IsNullOrWhiteSpace(bearer))
        {
            headers.Add("authorization", $"Bearer {bearer}");
        }
        var apiKey = _credentials.ApiKey;
        if (!string.IsNullOrWhiteSpace(apiKey))
        {
            headers.Add("x-api-key", apiKey);
        }
        return headers;
    }

    /// <summary>The shared credentials holder; mutating it hot-swaps creds on every client.</summary>
    public UdbCredentials Credentials => _credentials;

    internal GrpcChannel DataChannelForTesting => _dataChannel;

    internal GrpcChannel AuthChannelForTesting => _authChannel;

    internal GrpcChannel NativeServicesChannelForTesting => _nativeServicesChannel;

    /// <summary>
    /// Hot-swap the bearer token and API key sent on every subsequent call across
    /// all sub-clients. Call this after a login/refresh so the new token reaches
    /// outbound metadata — the C# analogue of the TypeScript <c>core.setCredentials</c>.
    /// </summary>
    public void SetCredentials(string? bearerToken, string? apiKey = null)
        => _credentials.Set(bearerToken, apiKey ?? _credentials.ApiKey);

    // ── first-class accessors ───────────────────────────────────────────────
    /// <summary>Raw data-plane broker client (Select/Upsert/etc.).</summary>
    public DataBroker.DataBrokerClient Data { get; }

    /// <summary>Authentication ergonomics (authenticate / refresh).</summary>
    public UdbAuthClient Auth { get; }

    /// <summary>Authorization ergonomics (can / require / batch / explain / native-access).</summary>
    public UdbAuthClient Authz { get; }

    /// <summary>Raw generated API-key service client.</summary>
    public ApikeyV1.ApiKeyService.ApiKeyServiceClient ApiKey { get; }

    /// <summary>Raw generated tenant service client.</summary>
    public TenantV1.TenantService.TenantServiceClient Tenant { get; }

    /// <summary>Raw generated notification service client.</summary>
    public NotificationV1.NotificationService.NotificationServiceClient Notification { get; }

    /// <summary>Raw generated analytics service client.</summary>
    public AnalyticsV1.AnalyticsService.AnalyticsServiceClient Analytics { get; }

    /// <summary>Storage ergonomics (upload/download/file metadata).</summary>
    public UdbStorageClient Storage { get; }

    /// <summary>Asset ergonomics (pipeline definitions, assets, step completion).</summary>
    public UdbAssetClient Asset { get; }

    /// <summary>WebRTC ergonomics, grouped: <c>Room</c> / <c>Peer</c> / <c>Track</c> / <c>Turn</c> / <c>Signaling</c>.</summary>
    public UdbWebRtcClient WebRtc { get; }

    // ── convenience wrappers ────────────────────────────────────────────────
    /// <summary>Run a SELECT through the data broker with the shared metadata.</summary>
    public Task<RecordSet> SelectAsync(SelectRequest request, CancellationToken ct = default)
        => Data.SelectAsync(request, Headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Run an UPSERT through the data broker with the shared metadata.</summary>
    public Task<MutationResponse> UpsertAsync(UpsertRequest request, CancellationToken ct = default)
        => Data.UpsertAsync(request, Headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>
    /// Send a notification. The tenant/project/correlation fields default from
    /// the facade config when left unset on the request.
    /// </summary>
    public Task<NotificationV1.SendNotificationResponse> SendNotificationAsync(
        NotificationV1.SendNotificationRequest request, CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(request.TenantId)) request.TenantId = _metadata.TenantId;
        if (string.IsNullOrEmpty(request.ProjectId)) request.ProjectId = _metadata.ProjectId;
        if (string.IsNullOrEmpty(request.CorrelationId)) request.CorrelationId = _metadata.CorrelationId;
        return Notification.SendNotificationAsync(request, Headers(), cancellationToken: ct).ResponseAsync;
    }

    /// <summary>Mint a new API key; returns the plaintext key alongside its metadata.</summary>
    public Task<ApikeyV1.CreateApiKeyResponse> CreateApiKeyAsync(
        ApikeyV1.CreateApiKeyRequest request, CancellationToken ct = default)
        => ApiKey.CreateApiKeyAsync(request, Headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Revoke an API key by id, with an optional reason.</summary>
    public Task<ApikeyV1.RevokeApiKeyResponse> RevokeApiKeyAsync(
        string keyId, string reason = "", CancellationToken ct = default)
        => ApiKey.RevokeApiKeyAsync(
            new ApikeyV1.RevokeApiKeyRequest { KeyId = keyId, RevokeReason = reason },
            Headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Create / onboard a tenant.</summary>
    public Task<TenantV1.CreateTenantResponse> CreateTenantAsync(
        TenantV1.CreateTenantRequest request, CancellationToken ct = default)
        => Tenant.CreateTenantAsync(request, Headers(), cancellationToken: ct).ResponseAsync;

    // ── lifecycle ───────────────────────────────────────────────────────────
    public ValueTask DisposeAsync()
    {
        Dispose();
        return ValueTask.CompletedTask;
    }

    public void Dispose()
    {
        _dataChannel.Dispose();
        if (_separateAuthChannel)
        {
            _authChannel.Dispose();
        }
        if (_separateWebrtcChannel)
        {
            _webrtcChannel.Dispose();
        }
    }
}
